### PR TITLE
imlib/draw: Add missing speculative read invalidates.

### PR DESCRIPTION
**Pull request by kwagyeman**
Fixes: https://github.com/openmv/openmv/issues/2219

### CODE REVIEW

File: src/omv/imlib/draw.c [hunk 1]
```diff
@@ -766,6 +766,10 @@ void imlib_draw_row_setup(imlib_draw_row_data_t *data) {
         data->dma2d.LayerCfg[1].ChromaSubSampling = DMA2D_NO_CSS;
         #endif
         HAL_DMA2D_ConfigLayer(&data->dma2d, 1);
+        #if __DCACHE_PRESENT
+        // SCB_InvalidateDCache_by_Addr does nothing is dsize is 0.
```
Comment (iabdalkader) [inline src/omv/imlib/draw.c:770]:
```suggestion
        // SCB_InvalidateDCache_by_Addr does nothing if dsize is 0.
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: src/omv/imlib/draw.c (hunk 1 above)
Comment (kwagyeman) [reply src/omv/imlib/draw.c]: Fixed

**Comment by kwagyeman** [comment]
I rechecked needing all of them. They are all required to follow the spec of invalidating cached reads to the DMA buffer.

File: src/omv/ports/stm32/jpeg.c [hunk 2]
```diff
@@ -744,6 +751,9 @@ void jpeg_decompress(image_t *dst, image_t *src) {
 
     if ((JPEG_state.jpeg_descr.Conf.ColorSpace == JPEG_YCBCR_COLORSPACE) && dst->is_color) {
         HAL_DMA2D_DeInit(&DMA2D_Handle);
+
+        // Ensure any cached reads are dropped.
+        SCB_InvalidateDCache_by_Addr((uint32_t *) dst->data, image_size(dst));
```
Comment (iabdalkader) [inline src/omv/ports/stm32/jpeg.c:756]: The whole dst image is invalidated right after configuring the DMA2D: https://github.com/openmv/openmv/blob/19d70d69e887cc64f67df653cc02f81db68be911/src/omv/ports/stm32/jpeg.c#L506-L510

Then every row/block is invalidated before the transfer: https://github.com/openmv/openmv/blob/19d70d69e887cc64f67df653cc02f81db68be911/src/omv/ports/stm32/jpeg.c#L712-L714

Now here `dst` is invalidated again after all transfers are complete.  The first invalidate is definitely Not needed, as every block will be invalidated anyway before the transfer, or all 3 of them can be replaced with just one invalidate after the `DMA2D_PollForTransfer`.

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: src/omv/ports/stm32/jpeg.c (hunk 2 above)
Comment (kwagyeman) [reply src/omv/ports/stm32/jpeg.c:756]: Yeah, that's why I dropped `SCB_InvalidateDCache_by_Addr((uint32_t *) rp, dst->w * dy * sizeof(uint16_t));` in the PR given the whole image is already invalidated before starting DMA2D.

> Also possible to replace all 3 of them with just one invalidate before/after the DMA2D_PollForTransfer.

Yeah, you could, there's less overhead doing them all in one go though before and after than before and after every row though right?

Comment (iabdalkader) [reply src/omv/ports/stm32/jpeg.c:756]: > before and after everything than before and after every row, right?

I think so.

Comment (kwagyeman) [reply src/omv/ports/stm32/jpeg.c:756]: Okay, then this PR is ready.

Comment (iabdalkader) [reply src/omv/ports/stm32/jpeg.c:756]: BTW if you want to save some time and optimize this, you could try invalidating the previous line/block *before* the poll.

Comment (iabdalkader) [reply src/omv/ports/stm32/jpeg.c:756]: That is invalidate while the DMA is transferring the next line/block.

Comment (kwagyeman) [reply src/omv/ports/stm32/jpeg.c]: Implemented the suggestion. Works.

**Review by iabdalkader — APPROVED** [review]

**Pull request merged**